### ISSUE TITLE
How can a rootless container (bind) mount directories or files from the host?

### ISSUE THREAD

**Issue by christian-2** [opencontainers/runc#3147]
I am looking for documentation or an example showing how rootless containers can mount directories and files from their host. The presentation [*Rootless containers with runC*](https://www.youtube.com/watch?v=r6EcUyamu94) from back in 2017 mentions (at 35:50) that "if you can access it as a regular user on a system ... it is literally just a matter of doing ... a bind mount". 

But is it really that simple? I am interested in employing runC in an rootless/rootless mode, i.e. where both `runc` and the container run rootless. Doesn't bind mount always require root privileges, so how could `runc` do a bind mount in that mode? So far I have created `config.json` with `runc spec --rootless ...` and defined some bind mounts under `mounts`. `runc create ...` then runs into errors. (Am I perhaps supposed to do the bind mount up front and outside `runc` as `root`?) 

My intended use case involves mounting e.g.`/etc/hostname`, or an SSH key, or say `/var/log/mystuff` from the host's filesystem into the rootless container. So again, is there some specific example or (better: and) documentation that shows (if this is indeed possible at all) how?

**Comment by cyphar** [comment]
Adding a bind-mount in mounts should work with rootless containers -- my first guess is that you didn't define a bind-mount correctly (you need to include `"bind"` in the list of options -- setting the type to "bind" isn't correct since mount types are ignored when you mount). The reason this works is that inside the user namespace you have privileges to mount.

**Comment by christian-2** [comment]
@cyphar Thanks for your quick response. I can by now mount a directory (that is owned by my user) from the host into the rootless container. Mount options include `bind`.

I am still struggling to mount e.g. `~/.ssh/id_rsa` (i.e. a file also owned by my user) or `ro`-mount `/etc/hostname` (i.e. a world-readable file owned by `root`). What specific entries in [`mounts`](https://github.com/opencontainers/runtime-spec/blob/master/config.md) could serve in those cases? My current assumption is that their mount points in rootfs would have to be [files](http://www.mardy.it/blog/2018/10/how-to-bind-mount-single-file.html) (also owned by my user from the host's perpective, i.e. by `root` from the container's.)

One reason I'd like to do this with rootless runC is that it available as Debian package (e.g. `runc=1.0.0~rc6+dfsg1-3` in Buster). This could provide a chance to avoid outside dependencies for my workload. (It has run under Docker so far.)

**Comment by cyphar** [comment]
Can you post the config you're using and the error you're getting? Mounting over files should work without issue (and the original owner of the source of the mountpoint doesn't affect bind-mounting).

**Comment by AkihiroSuda** [comment]
> runc=1.0.0~rc6+dfsg1-3

That package is too old.  Please use v1.0.1.

(Included in `containerd.io` dpkg maintained by Docker, Inc. https://download.docker.com/linux/debian/dists/buster/pool/stable/amd64/)

**Comment by christian-2** [comment]
I have now upgraded to Debian 11.0 (with `kernel 5.10.0-8-amd64`). Since its latest version is still only `runc=1.0.0~rc93+ds1-5+b2`, I have installed runc v1.0.1 separately as part of Debian's `containerd.io=1.4.9-1`, as was suggested. I shall switch back to the Debian package once its version has caught up.

With that, all intended bind mounts into the rootless container now apparently do work. I'll close the issue for now. Thanks again @AkihiroSuda and @cyphar.
